File: summary/go/14_greedy.go
package main

import (
	"fmt"
	"sort"
)

type Item struct {
	Value  float64
	Weight float64
	Ratio  float64
	Index  int
}

func greedyAlgorithm(items []Item, capacity float64) (float64, []int) {
	sort.Slice(items, func(i, j int) bool {
		return items[i].Ratio > items[j].Ratio
	})

	totalValue := 0.0
	totalWeight := 0.0
	selectedItems := []int{}

	for _, item := range items {
		if totalWeight+item.Weight <= capacity {
			selectedItems = append(selectedItems, item.Index)
			totalValue += item.Value
			totalWeight += item.Weight
		}
	}

	return totalValue, selectedItems
}

// Example usage
func main() {
	items := []Item{
		{Value: 60, Weight: 10, Ratio: 6, Index: 1},
		{Value: 100, Weight: 20, Ratio: 5, Index: 2},
		{Value: 120, Weight: 30, Ratio: 4, Index: 3},
	}
	capacity := 50.0

	totalValue, selectedItems := greedyAlgorithm(items, capacity)
	fmt.Printf("Total value: %.2f\n", totalValue)
	fmt.Println("Items selected:", selectedItems)
}

File: summary/ts/14_greedy.ts
function greedyAlgorithm(values: number[], weights: number[], capacity: number): [number, number[]] {
    const items = values.map((value, index) => ({
        index: index + 1,
        value: value,
        weight: weights[index],
        ratio: value / weights[index]
    }));
    items.sort((a, b) => b.ratio - a.ratio);

    let totalValue = 0;
    let totalWeight = 0;
    const selectedItems: number[] = [];

    items.forEach(item => {
        if (totalWeight + item.weight <= capacity) {
            selectedItems.push(item.index);
            totalValue += item.value;
            totalWeight += item.weight;
        }
    });

    return [totalValue, selectedItems];
}

// Example usage:
const values = [60, 100, 120];
const weights = [10, 20, 30];
const capacity = 50;
const [value, items] = greedyAlgorithm(values, weights, capacity);
console.log(`Total value: ${value}`);
console.log(`Items selected: ${items.join(', ')}`);
